contracts/OnChainVault.sol

```diff
@@ -0,0 +1,425 @@
+import { ERC20Upgradeable } from "@openzeppelin/contracts-upgradeable/token/ERC20/ERC20Upgradeable.sol";
+import { OwnableUpgradeable } from "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+
+import { StrategyParams, IOnChainVault } from "./interfaces/IOnChainVault.sol";
+import { IBaseStrategy } from "./interfaces/IBaseStrategy.sol";
+
+contract OnChainVault is
+    Initializable,
+    ERC20Upgradeable,
+    IOnChainVault,
+    OwnableUpgradeable
+{
+    using SafeERC20 for IERC20;
+    using SafeERC20 for ERC20;
+
+    function initialize(
+        IERC20 _token,
+        address _governance,
+        address treasury,
+        string calldata name,
+        string calldata symbol
+    ) external initializer {
+        __Ownable_init();
+        __ERC20_init(name, symbol);
+
+        governance = _governance;
+        token = _token;
+        approve(treasury, type(uint256).max);
+    }
+
+    uint256 public constant MAX_BPS = 10_000;
+
+    address public override governance;
+    IERC20 public override token;
+    uint256 public depositLimit;
+    uint256 public totalDebtRatio;
+    uint256 public totalDebt;
+    bool public emergencyShutdown;
+    mapping(address => StrategyParams) public strategies;
+    mapping(address strategy => uint256 position)
+        public strategyPositionInArray;
+
+    address[] public OnChainStrategies;
+
+    modifier onlyAuthorized() {
+        if (msg.sender != governance || msg.sender != owner())
+            revert Vault__OnlyAuthorized(msg.sender);
+        _;
+    }
+
+    function decimals() public view virtual override returns (uint8) {
+        return ERC20(address(token)).decimals();
+    }
+
+    function revokeFunds() external onlyAuthorized {
+        payable(msg.sender).transfer(address(this).balance);
+    }
+
+    function setEmergencyShutdown(
+        bool _emergencyShutdown
+    ) external onlyAuthorized {
+        emergencyShutdown = _emergencyShutdown;
+    }
+
+    function setDepositLimit(uint256 _limit) external onlyAuthorized {
+        depositLimit = _limit;
+    }
+    //!Tests are not working with this implimentation of PPS
+    //!TODO: rework system to exlude dependencies of totalDebt, need to rethink logic of vault, big work
+    function totalAssets() public view returns (uint256 _assets) {
+        for (uint256 i = 0; i < OnChainStrategies.length; i++) {
+            _assets += IBaseStrategy(OnChainStrategies[i])
+                .estimatedTotalAssets();
+        }
+        // _assets += totalIdle() + totalDebt;
+    }
+
+    function totalIdle() public view returns (uint256) {
+        return token.balanceOf(address(this));
+    }
+
+    function deposit(
+        uint256 _amount,
+        address _recipient
+    ) external returns (uint256) {
+        if (_amount > depositLimit) revert Vault__DepositLimit();
+        return _deposit(_amount, _recipient);
+    }
+
+    function deposit(uint256 _amount) external returns (uint256) {
```

Comment:
Duplication of code. Suggestion: make `deposit(uint256, address)` visibility to `public` and then reuse it `function deposit(uint256 _amount) external returns (uint256) {
    deposit(_amount, msg.sender);
}`.

Reply:
there will be more jumps opcodes => more gas consumption